lmdeploy/serve/turbomind/deploy.py

```diff
@@ -379,172 +143,892 @@ def permute(x: torch.Tensor):
+                                param_name,
+                                [param_data.size(0), size * n_ckpt])
+                            param.data[:, size * i:size * (i + 1)] = param_data
+                        else:  # bias
+                            param = get_param(param_name, [size])
+                            param.data = param_data
+                    elif i == 0:
+                        param = get_param(param_name, param_data.size())
+                        param.data = param_data
+                del ckpt
+
+            for name, param in model_params.items():
+                # transpose all weights as TurboMind is expecting column-major
+                # weights:
+                # (output_dims, input_dims) -> (input_dims, output_dims)
+                key = name.split('.')[-2]
+                if key in ['w1', 'w3', 'wq', 'wk', 'wv', 'w2', 'wo']:
+                    param.data = param.data.t()
+                    if key in ['wq', 'wk']:
+                        param.data = reverse_permute(param.data)
+
+            # concat qkv projection
+            for t in ['weight', 'bias']:
+                for i in range(self.num_layer):
+                    # attn
+                    try:
+                        _tm_keys = ['q', 'k', 'v', 'o']
+                        _fb_keys = ['wq', 'wk', 'wv', 'wo']
+                        for tm_key, fb_key in zip(_tm_keys, _fb_keys):
+                            params[i][f'self_attn.{tm_key}.{t}'] = \
+                                model_params.pop(
+                                    f'layers.{i}.attention.{fb_key}.{t}')
+                    except:  # noqa: E722
+                        pass
+
+                    if t == 'bias':
+                        continue
+                    # ffn
+                    for key in ['w1', 'w2', 'w3']:
+                        params[i][f'ffn.{key}.weight'] = \
+                            model_params.pop(
+                                f'layers.{i}.feed_forward.{key}.{t}')
+                    # norm
+                    params[i]['attn_norm.weight'] = \
+                        model_params.pop(f'layers.{i}.attention_norm.{t}')
+                    params[i]['ffn_norm.weight'] = \
+                        model_params.pop(f'layers.{i}.ffn_norm.{t}')
+
+            others = ['tok_embeddings.weight', 'norm.weight', 'output.weight']
+            for key in others:
+                params[key] = model_params.pop(key)
+            self.params = params
+
+        def layers(self):
+            for layer_id in range(self.num_layer):
+                yield LlamaDecoderLayer(layer_id, self.params[layer_id])
+
+    def bins(self):
+        for bin in self.bin_files:
+            yield bin
+
+
+class HfModel:
+
+    def __init__(self, model_path: str, tokenizer_path: str, **kwargs):
+        self.model_path = model_path
+        self.tokenizer_path = tokenizer_path
+
+        if 'quant_path' in kwargs and kwargs['quant_path'] is not None:
```

Comment:
I don't think `HfModel` should tackle `quant_path`. It is an input for `awq` model.